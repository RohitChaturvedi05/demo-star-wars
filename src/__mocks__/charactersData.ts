import {
    CharacterDetails,
    Characters,
    Film,
    Starship,
} from '../models/Character';

export const mockCharacters: Characters = [
    {
        uid: '1',
        id: '1',
        name: 'Luke Skywalker',
        gender: 'male',
        skin_color: 'fair',
        hair_color: 'blond',
        height: '172',
        eye_color: 'blue',
        mass: '77',
        birth_year: '19BBY',
        homeworldUrl: 'http://swapi.dev/api/planets/1/',
        homeworld: '1',
    },
    {
        uid: '2',
        id: '2',
        name: 'Leia Organa',
        gender: 'female',
        skin_color: 'light',
        hair_color: 'brown',
        height: '150',
        eye_color: 'brown',
        mass: '49',
        birth_year: '19BBY',
        homeworldUrl: 'http://swapi.dev/api/planets/2/',
        homeworld: '2',
    },
];

export const mockFavoriteCharacters: Characters = [
    {
        uid: '1',
        id: '1',
        name: 'Luke Skywalker',
        gender: 'male',
        skin_color: 'fair',
        hair_color: 'blond',
        height: '172',
        eye_color: 'blue',
        mass: '77',
        birth_year: '19BBY',
        homeworldUrl: 'http://swapi.dev/api/planets/1/',
        homeworld: '1',
    },
];

export const mockPlanets = {
    1: {
        id: '1',
        name: 'Tatooine',
        diameter: '10465',
        gravity: '1',
        population: '200000',
    },
    2: {
        id: '2',
        name: 'Alderaan',
        diameter: '12500',
        gravity: '1',
        population: '2000000000',
    },
};

export const mockCharacterDetails: CharacterDetails = {
    uid: '1',
    id: '1',
    name: 'Luke Skywalker',
    gender: 'male',
    skin_color: 'fair',
    hair_color: 'blond',
    height: '172',
    eye_color: 'blue',
    mass: '77',
    birth_year: '19BBY',
    homeworldUrl: 'http://swapi.dev/api/planets/1/',
    homeworld: '1',
    films: ['http://swapi.dev/api/film/1', 'http://swapi.dev/api/film/2'],
    starships: [
        'http://swapi.dev/api/starship/1',
        'http://swapi.dev/api/starship/2',
    ],
};

export const mockFilms: Film[] = [
    {
        id: '1',
        title: 'A New Hope',
        director: 'George Lucas',
        producer: 'Gary Kurtz, Rick McCallum',
        release_date: '1977-05-25',
    },
    {
        id: '2',
        title: 'The Empire Strikes Back',
        director: 'Irvin Kershner',
        producer: 'Gary Kurtz, Rick McCallum',
        release_date: '1980-05-17',
    },
];

export const mockStarship: Starship[] = [
    {
        name: 'Millennium Falcon',
        cargo_capacity: '100000',
        passengers: '6',
        max_atmosphering_speed: '1050',
        manufacturer: 'Corellian Engineering Corporation',
    },
    {
        name: 'X-wing',
        cargo_capacity: '100000',
        passengers: '6',
        max_atmosphering_speed: '1050',
        manufacturer: 'Incom Corporation',
    },
];
