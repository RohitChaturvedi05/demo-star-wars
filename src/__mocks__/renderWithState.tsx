import { render } from '@testing-library/react';
import React from 'react';
import { StateProvider } from '../state';

export const renderWithState = (Component: React.ReactNode) => {
    return render(<StateProvider>{Component}</StateProvider>);
};
