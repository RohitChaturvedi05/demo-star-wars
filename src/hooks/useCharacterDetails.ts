import { useCallback, useMemo, useState } from 'react';
import { Gender } from '../components/Gender';
import { Character, Film, Starship } from '../models/Character';
import { getCharacterById } from '../services/getCharacterById';
import { getFilm } from '../services/getFilms';
import { getPlanetsByIds } from '../services/getPlanets';
import { getStarShip } from '../services/getStarship';
import { useStateContext } from '../state';
import { ActionType } from '../state/actions';

export const useCharacterDetails = () => {
    const [message, setMessage] = useState('');
    const { state, dispatch } = useStateContext();
    const { planets } = state;
    const [character, setCharacter] = useState<Character>();
    const [films, setFilms] = useState<Film[]>([]);
    const [starShips, setStarShips] = useState<Starship[]>([]);

    const getCharacterDetails = useCallback(
        async (id: string) => {
            const characterData = await getCharacterById(id);
            if (characterData) {
                const planetId = parseInt(characterData?.homeworld ?? '');
                if (planets[planetId] === undefined) {
                    const data = await getPlanetsByIds([planetId]);
                    dispatch({ type: ActionType.SET_PLANETS, payload: data });
                }
                const filePromise = characterData?.films.map(getFilm);
                const starShipPromise =
                    characterData?.starships.map(getStarShip);

                const allFilms = await Promise.all(filePromise);
                const allStarShips = await Promise.all(starShipPromise);

                setFilms(allFilms.filter((film) => !!film) as Film[]);
                setStarShips(
                    allStarShips.filter((starShip) => !!starShip) as Starship[]
                );
                setCharacter(characterData);
            }
        },
        [dispatch, planets]
    );

    const onAddToFavorites = useCallback(
        (id: string) => {
            dispatch({
                type: ActionType.SET_FAV_CHARACTERS,
                payload: id,
            });
        },
        [dispatch]
    );

    const filmsNames = useMemo(
        () => films.map((film) => ({ name: film.title })),
        [films]
    );
    const starShipsNames = useMemo(
        () => starShips.map((starShip) => ({ name: starShip.name })),
        [starShips]
    );

    const isFavorite = useMemo(
        () => character && state.favCharacters.includes(character.id),
        [character, state.favCharacters]
    );

    const onGenderChange = (value: Gender, id: string) => {
        dispatch({ type: ActionType.SET_GENDER, payload: { value, id } });
        setTimeout(() => {
            setMessage('Gender Saved !!');
        }, 2000);
    };

    return {
        message,
        films,
        starShips,
        planets,
        character,
        filmsNames,
        isFavorite,
        starShipsNames,
        onGenderChange,
        onAddToFavorites,
        getCharacterDetails,
    };
};
