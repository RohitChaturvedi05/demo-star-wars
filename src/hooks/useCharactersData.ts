import { useCallback, useMemo, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { APP_PATH } from '../routes/constants';
import { getCharacters } from '../services/getCharacters';
import { getPlanetsByIds } from '../services/getPlanets';
import { getCharactersSearch } from '../services/searchByCharacterName';
import { extractPlanetId } from '../services/utils/extractPlanetId';
import { useStateContext } from '../state';
import { ActionType } from '../state/actions';
import { useDebounce } from './useDebounce';

export const useCharactersData = () => {
    const [queryTerm, setQueryTerm] = useState('');
    const debounceQuery = useDebounce(queryTerm, 300);

    const [isLoading, setLoading] = useState(false);
    const { state, dispatch } = useStateContext();
    const { planets } = state;

    const navigate = useNavigate();

    const getCharactersData = useCallback(
        async (page: number) => {
            setLoading(true);
            const { characters, total_pages } = await getCharacters({
                page,
            });

            // get planets ids from characters
            const planetsId = characters
                .map((item) => extractPlanetId(item.homeworldUrl))
                .filter((id): id is number => id !== null && id !== undefined)
                .filter((id) => planets[id] === undefined);

            // Update planets state if needed
            if (planetsId.length > 0) {
                const planetsData = await getPlanetsByIds(
                    Array.from(new Set(planetsId))
                );
                dispatch({
                    type: ActionType.SET_PLANETS,
                    payload: planetsData,
                });
            }
            dispatch({ type: ActionType.SET_CHARACTERS, payload: characters });
            dispatch({ type: ActionType.SET_TOTAL_PAGE, payload: total_pages });
            dispatch({
                type: ActionType.SET_CURRENT_PAGE,
                payload: page,
            });

            setLoading(false);
        },
        [dispatch, planets]
    );

    const searchCharactersByName = useCallback(
        async (name: string) => {
            setLoading(true);
            const { characters } = await getCharactersSearch({
                page: 1,
                name,
            });
            dispatch({ type: ActionType.SET_CHARACTERS, payload: characters });
            dispatch({ type: ActionType.SET_TOTAL_PAGE, payload: 0 });
            setLoading(false);
        },
        [dispatch]
    );

    const onFavoriteToggle = useCallback(
        (id: string) => {
            dispatch({
                type: ActionType.SET_FAV_CHARACTERS,
                payload: id,
            });
        },
        [dispatch]
    );

    const onDetailClick = useCallback(
        (id: string) => {
            navigate(`${APP_PATH.DETAILS}/${id}`);
        },
        [navigate]
    );

    const onPageChange = async (page: number) => {
        dispatch({
            type: ActionType.SET_CURRENT_PAGE,
            payload: page,
        });
        await getCharactersData(page);
    };

    const onSearch = async (searchTerm: string) => {
        setQueryTerm(searchTerm);
        setLoading(true);
        if (searchTerm.length === 0) {
            await onPageChange(1);
        }
    };

    const initializeState = useCallback(
        () => getCharactersData(1),
        [getCharactersData]
    );

    const favorites = useMemo(() => {
        return state.characters.filter(({ uid }) =>
            state.favCharacters.includes(uid)
        );
    }, [state.characters, state.favCharacters]);

    return {
        isLoading,
        favorites,
        queryTerm,
        debounceQuery,
        characters: state.characters,
        planets: state.planets,
        favCharacters: state.favCharacters,
        currentPage: state.currentPage,
        totalPage: state.totalPage,
        initializeState,
        onSearch,
        searchCharactersByName,
        onFavoriteToggle,
        onDetailClick,
        getCharactersData,
        onPageChange,
    };
};
