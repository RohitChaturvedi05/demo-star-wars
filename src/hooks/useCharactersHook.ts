import { useCallback } from 'react';
import { useNavigate } from 'react-router-dom';
import { APP_PATH } from '../routes/constants';
import { useStateContext } from '../state';
import { ActionType } from '../state/actions';

export const useCharactersHook = () => {
    const { dispatch } = useStateContext();
    const navigate = useNavigate();

    const onFavoriteToggle = useCallback(
        (id: string) => {
            dispatch({
                type: ActionType.SET_FAV_CHARACTERS,
                payload: id,
            });
        },
        [dispatch]
    );

    const onDetailClick = useCallback(
        (id: string) => {
            navigate(`${APP_PATH.DETAILS}/${id}`);
        },
        [navigate]
    );

    return {
        onFavoriteToggle,
        onDetailClick,
    };
};
