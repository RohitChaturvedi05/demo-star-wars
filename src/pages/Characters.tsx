import { Box, Toolbar, Typography } from '@mui/material';
import { useEffect } from 'react';
import { styled } from 'styled-components';
import { CharacterList } from '../components/CharacterList';
import { Search } from '../components/Search';
import { useCharactersData } from '../hooks/useCharactersData';

const PageContainer = styled.div`
    padding: 24px;
    width: 100%;
`;
const ResponsiveBox = styled(Box)`
    display: flex;
    justify-content: space-between;

    @media (max-width: 600px) {
        flex-direction: column;
        align-items: flex-start;
    }
`;

export const Characters = () => {
    const {
        debounceQuery,
        isLoading,
        characters,
        currentPage,
        favCharacters,
        totalPage,
        planets,
        onSearch,
        initializeState,
        searchCharactersByName,
        onPageChange,
        onDetailClick,
        onFavoriteToggle,
    } = useCharactersData();

    useEffect(() => {
        (async () => {
            await initializeState();
        })();
        // need to initialize the state only once at the mount
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    useEffect(() => {
        (async () => {
            if (debounceQuery.length > 0) {
                await searchCharactersByName(debounceQuery);
            }
        })();
    }, [debounceQuery, searchCharactersByName]);

    return (
        <>
            <Toolbar />

            <PageContainer>
                <ResponsiveBox>
                    <Typography variant="h4" component="h1">
                        Star Wars Characters
                    </Typography>
                    <Search onSearch={onSearch} />
                </ResponsiveBox>

                <CharacterList
                    isLoading={isLoading}
                    planets={planets}
                    characters={characters}
                    favCharacters={favCharacters}
                    pagination={{
                        currentPage,
                        totalPage,
                    }}
                    showPagination
                    onFavoriteToggle={onFavoriteToggle}
                    onDetailClick={onDetailClick}
                    onPageChange={onPageChange}
                />
            </PageContainer>
        </>
    );
};
