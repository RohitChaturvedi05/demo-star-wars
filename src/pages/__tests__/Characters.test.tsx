import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { mockCharacters, mockPlanets } from '../../__mocks__/charactersData';
import { useCharactersData } from '../../hooks/useCharactersData';
import { Characters } from '../Characters';

// Acceptance Criteria - Character list view
// o list all characters from the Star Wars universe (name, gender & home planet)
// o provide pagination controls to step through the results
// o provide a search field to query by character name
// o clicking a list entry should navigate to the character details page

jest.mock('react-router-dom', () => ({
    useNavigate: jest.fn(),
}));
jest.mock('../../hooks/useCharactersData');

const MockUseCharactersData = {
    isLoading: false,
    characters: mockCharacters,
    planets: mockPlanets,
    currentPage: 1,
    totalPage: 3,
    favCharacters: [],
    debounceQuery: '',
    queryTerm: '',
    onSearch: jest.fn(),
    initializeState: jest.fn(),
    searchCharactersByName: jest.fn(),
    onPageChange: jest.fn(),
    getCharactersData: jest.fn(),
    onDetailClick: jest.fn(),
    onFavoriteToggle: jest.fn(),
};

describe('pages/Characters', () => {
    beforeEach(() => {
        (useCharactersData as jest.Mock).mockReturnValue(MockUseCharactersData);
    });

    it('should render without crashing', () => {
        const { container } = render(<Characters />);
        expect(container).toMatchSnapshot();
    });

    it('should render list of characters with name, gender and home planet', () => {
        render(<Characters />);

        expect(screen.getByText('Luke Skywalker')).toBeInTheDocument();
        expect(screen.getByText('male')).toBeInTheDocument();
        expect(screen.getByText('Tatooine')).toBeInTheDocument();

        expect(screen.getByText('Leia Organa')).toBeInTheDocument();
        expect(screen.getByText('female')).toBeInTheDocument();
        expect(screen.getByText('Alderaan')).toBeInTheDocument();
    });

    it('should show pagination controls', () => {
        render(<Characters />);

        expect(screen.getByRole('navigation')).toBeInTheDocument();
        const paginationButtons = screen.getAllByRole('button');
        expect(paginationButtons.length).toBeGreaterThan(0);
    });

    it('should handle pagination click', async () => {
        const { onPageChange } = useCharactersData();
        render(<Characters />);

        const nextPageButton = screen.getByRole('button', {
            name: /go to page 2/i,
        });
        fireEvent.click(nextPageButton);

        expect(onPageChange).toHaveBeenCalledWith(2);
    });

    it('should provide search functionality', async () => {
        const { onSearch } = useCharactersData();
        render(<Characters />);

        const searchInput = screen.getByRole('textbox');
        fireEvent.change(searchInput, { target: { value: 'Luke' } });

        await waitFor(
            () => {
                expect(onSearch).toHaveBeenCalledWith('Luke');
            },
            { timeout: 400 }
        );
    });

    it('should handle character detail navigation', () => {
        render(<Characters />);
        const uid = '1';
        const detailsBtn = screen.getByTestId(`btn-details-${uid}`);
        fireEvent.click(detailsBtn);
        expect(MockUseCharactersData.onDetailClick).toHaveBeenCalledWith(uid);
    });

    it('should show loading state', () => {
        (useCharactersData as jest.Mock).mockReturnValue({
            ...MockUseCharactersData,
            isLoading: true,
        });

        render(<Characters />);
        expect(screen.getByTestId('loader')).toBeInTheDocument();
    });
});
