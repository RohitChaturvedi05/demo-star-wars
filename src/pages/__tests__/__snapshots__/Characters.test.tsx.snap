// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/Characters should render without crashing 1`] = `
<div>
  <div
    class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1ygil4i-MuiToolbar-root"
  />
  <div
    class="sc-FEMpB jnBMmq"
  >
    <div
      class="sc-gjcoXW bTDTZW MuiBox-root css-0"
    >
      <h1
        class="MuiTypography-root MuiTypography-h4 css-f84ov1-MuiTypography-root"
      >
        Star Wars Characters
      </h1>
      <form
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-70tf50-MuiPaper-root"
        style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
      >
        <div
          class="MuiInputBase-root MuiInputBase-colorPrimary css-932rvu-MuiInputBase-root"
          data-testid="search-box"
        >
          <input
            aria-label="search"
            class="MuiInputBase-input css-yimnyd-MuiInputBase-input"
            placeholder="Search"
            type="text"
          />
        </div>
        <button
          aria-label="search"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-d49qdc-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
            data-testid="SearchIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
            />
          </svg>
        </button>
      </form>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-y2ff7i-MuiPaper-root-MuiTableContainer-root"
      data-testid="character-list"
      style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
    >
      <table
        class="MuiTable-root css-1xwxv7r-MuiTable-root"
        data-testid="character-list$table"
      >
        <thead
          class="MuiTableHead-root css-1a7iywq-MuiTableHead-root"
          data-testid="character-list$head"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-1xtozoh-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1orzuox-MuiTableCell-root"
              scope="col"
            />
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1orzuox-MuiTableCell-root"
              scope="col"
            >
              Name
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1orzuox-MuiTableCell-root"
              scope="col"
            >
              Gender
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1orzuox-MuiTableCell-root"
              scope="col"
            >
              Home planet
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1orzuox-MuiTableCell-root"
              scope="col"
            />
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-gmh7jj-MuiTableBody-root"
          data-testid="character-list$body"
        >
          <tr
            class="MuiTableRow-root css-1xtozoh-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-53g0n7-MuiButtonBase-root-MuiIconButton-root"
                data-testid="btn-fav-1"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                  data-testid="StarBorderOutlinedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="m22 9.24-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28z"
                  />
                </svg>
              </button>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
            >
              Luke Skywalker
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
            >
              male
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
            >
              Tatooine
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-53g0n7-MuiButtonBase-root-MuiIconButton-root"
                data-testid="btn-details-1"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                  data-testid="ListAltIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 5v14H5V5zm1.1-2H3.9c-.5 0-.9.4-.9.9v16.2c0 .4.4.9.9.9h16.2c.4 0 .9-.5.9-.9V3.9c0-.5-.5-.9-.9-.9M11 7h6v2h-6zm0 4h6v2h-6zm0 4h6v2h-6zM7 7h2v2H7zm0 4h2v2H7zm0 4h2v2H7z"
                  />
                </svg>
              </button>
            </td>
          </tr>
          <tr
            class="MuiTableRow-root css-1xtozoh-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-53g0n7-MuiButtonBase-root-MuiIconButton-root"
                data-testid="btn-fav-2"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                  data-testid="StarBorderOutlinedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="m22 9.24-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28z"
                  />
                </svg>
              </button>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
            >
              Leia Organa
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
            >
              female
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
            >
              Alderaan
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-53g0n7-MuiButtonBase-root-MuiIconButton-root"
                data-testid="btn-details-2"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                  data-testid="ListAltIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 5v14H5V5zm1.1-2H3.9c-.5 0-.9.4-.9.9v16.2c0 .4.4.9.9.9h16.2c.4 0 .9-.5.9-.9V3.9c0-.5-.5-.9-.9-.9M11 7h6v2h-6zm0 4h6v2h-6zm0 4h6v2h-6zM7 7h2v2H7zm0 4h2v2H7zm0 4h2v2H7z"
                  />
                </svg>
              </button>
            </td>
          </tr>
        </tbody>
      </table>
      <div
        class="MuiBox-root css-1qc63lq"
      >
        <nav
          aria-label="pagination navigation"
          class="MuiPagination-root MuiPagination-text css-1xra8g6-MuiPagination-root"
          data-testid="character-list$pagination"
        >
          <ul
            class="MuiPagination-ul css-1c5o7vy-MuiPagination-ul"
          >
            <li>
              <button
                aria-label="Go to previous page"
                class="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-rounded Mui-disabled MuiPaginationItem-previousNext css-6qpcc8-MuiButtonBase-root-MuiPaginationItem-root"
                disabled=""
                tabindex="-1"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-8uzxat-MuiSvgIcon-root-MuiPaginationItem-icon"
                  data-testid="NavigateBeforeIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                  />
                </svg>
              </button>
            </li>
            <li>
              <button
                aria-current="page"
                aria-label="page 1"
                class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-rounded Mui-selected MuiPaginationItem-page css-6qpcc8-MuiButtonBase-root-MuiPaginationItem-root"
                tabindex="0"
                type="button"
              >
                1
              </button>
            </li>
            <li>
              <button
                aria-label="Go to page 2"
                class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-rounded MuiPaginationItem-page css-6qpcc8-MuiButtonBase-root-MuiPaginationItem-root"
                tabindex="0"
                type="button"
              >
                2
              </button>
            </li>
            <li>
              <button
                aria-label="Go to page 3"
                class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-rounded MuiPaginationItem-page css-6qpcc8-MuiButtonBase-root-MuiPaginationItem-root"
                tabindex="0"
                type="button"
              >
                3
              </button>
            </li>
            <li>
              <button
                aria-label="Go to next page"
                class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-rounded MuiPaginationItem-previousNext css-6qpcc8-MuiButtonBase-root-MuiPaginationItem-root"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-8uzxat-MuiSvgIcon-root-MuiPaginationItem-icon"
                  data-testid="NavigateNextIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                  />
                </svg>
              </button>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </div>
</div>
`;
